### DISCUSSION TITLE
Allow to customize esbuild (plugins, pre-transformation)

### DISCUSSION THREAD

**Discussion by tobiasdiez** [nuxt/nuxt#15905] [Ideas]
Some libraries or typescript features are not (yet) compatible with esbuild. For example, `emitDecoratorMetadata` is not supported by esbuild and thus libraries using typescript operators are not working (like dependency injection, TypeORM or typed graphql, all pretty popular on the backend).

Often there are workarounds for these problems, which involve one or more of the following:
- Transform the code before it is passed to esbuild
- Run esbuild plugins
- Replace esbuild completely, say by swc

Continuing the metadata example, one could a) pre-compile the code with `tsc` before passing it to esbuild b) use plugins such as [esbuild-plugin-tsc](https://github.com/thomaschaaf/esbuild-plugin-tsc), [esbuild-decorators](https://github.com/anatine/esbuildnx) or c) [run swc in place of esbuild](https://stackoverflow.com/a/68570844/873661) for the typescript files that use decorators. As of now, nuxt 3 and bridge doesn't seem to provide any options to allow any of these things.

**Comment by danielroe** [comment]
You can set esbuild options with:
```js
export default defineNuxtConfig({
  nitro: {
    esbuild: {
      options: {
        // options from https://github.com/egoist/rollup-plugin-esbuild
      }
    }
  }
})
```

**Reply by tobiasdiez** [reply]
Thanks the suggestion. This might be enough for some use cases. However, since https://github.com/nuxt/framework/blob/ddf45fbcc7b707a50e0af55740a934f8e99e07e7/packages/nitro/src/rollup/plugins/esbuild.ts#L85
uses `transform` instead of `build` from esbuild, plugins don't work (this is by design, according to the esbuild author). I also don't see any possibility in the nuxt code to transform `code` before it is passed to esbuild.

**Reply by pi0** [reply]
Indeed and it is the same limitation we have with vite (https://github.com/vitejs/vite/issues/2779). Depending on esbuild plugins in rollup setups seems not a viable solution at the moment.

> Transform the code before it is passed to esbuild

This seems the best option to move forward. We can use [`nitro:rollup:before` hook](https://github.com/nuxt/framework/blob/main/packages/nitro/src/build.ts#L57) to add transform plugins.

**Reply by tobiasdiez** [reply]
Thanks a lot for the pointer. Adding the following configuration indeed embeds the metadada:
```ts
import typescript from '@rollup/plugin-typescript'
....

nitro: {
    hooks: {
      'nitro:rollup:before'(ctx) {
        ctx.rollupConfig?.plugins?.unshift(typescript())
      },
    },
  },
``` 
However, there is still a problem that imports that are unused but have sideeffects are deleted (that's already haping without the typescript plugin added). So for example the `import 'reflect-metadata'` is just deleted (the same happens also with `import './myfile.ts'`). Adding it manually to the generated `nitro/index.mjs` and everything works. Is there an option to preserve those imports?

**Reply by tobiasdiez** [reply]
Update: adding `reflect-metadata` to `moduleSideEffects` solves the last issue as well.

**Reply by mrshmllow** [reply]
using
```
nitro: {
  hooks: {
    'nitro:rollup:before'(ctx) {
      ctx.moduleSideEffects.push('reflect-metadata')
    }
  }
}
```
it started to include `import 'reflect-metadata';` in nitro/index.mjs

**Reply by ahoiroman** [reply]
Unfortunately this does not seem to work any longer. I added it, but it's not included.

**Reply by RealTirex** [reply]
@ahoiroman you can still use:
```javascript
    nitro: {
        hooks: {
            "rollup:before"(ctx) {
                ctx.options.moduleSideEffects.push("reflect-metadata")
            }
        }
    }
```

**Reply by productdevbook** [reply]
https://github.com/abraham/reflection

reflect-metadata alternative has esm support, I recommend it
